tests/Feature/Livewire/CourseListTest.php

```diff
@@ -44,11 +45,116 @@
+        ->has(Episode::factory())
+        ->has(
+            Tag::factory()
+                ->count(2)
+                ->state(new Sequence(
+                    ['name' => 'Bootstrap'],
+                    ['name' => 'AlpineJS'],
+                ))
+        )
+        ->create();
+
+     Course::factory()
+        ->has(Episode::factory())
+        ->has(
+            Tag::factory()
+                ->count(2)
+                ->state(new Sequence(
+                    ['name' => 'Vue'],
+                    ['name' => 'Tailwind'],
+                ))
+        )
+        ->create();
+
+    $firstCourseTagsIds = $filament->tags()->pluck('tag_id')->toArray();
+
+    Livewire::test(CourseList::class)
+        ->assertCanSeeTableRecords(Course::get())
+        ->filterTable('tags',$firstCourseTagsIds)
+        ->assertCanSeeTableRecords(Course::whereHas('tags',function ($q) use ($firstCourseTagsIds){
+            $q->whereIn('tag_id',$firstCourseTagsIds);
+        })->get())
+        ->assertCanNotSeeTableRecords(Course::whereHas('tags',function ($q) use ($firstCourseTagsIds){
+            $q->whereNotIn('tag_id',$firstCourseTagsIds);
+        })->get())
+        ->assertSeeText([
+            'Bootstrap','AlpineJS'
+        ])
+        ->assertDontSeeText([
+            'Tailwind','Vue'
+        ]);
+
+});
+
+
+it('can be filter remove tags', function () {
+
+    $filament = Course::factory()
+        ->has(Episode::factory())
+        ->has(
+            Tag::factory()
+                ->count(2)
+                ->state(new Sequence(
+                    ['name' => 'Bootstrap'],
+                    ['name' => 'AlpineJS'],
+                ))
+        )
+        ->create();
+
+    $vues = Course::factory()
+        ->has(Episode::factory())
+        ->has(
+            Tag::factory()
+                ->count(2)
+                ->state(new Sequence(
+                    ['name' => 'Vue'],
+                ))
+        )
+        ->create();
+
+    $filamentCourseTagsIds = $filament->tags()->pluck('tag_id')->toArray();
+    $vuesCourseTagsIds = $vues->tags()->pluck('tag_id')->toArray();
+
+    Livewire::test(CourseList::class)
+        ->assertCanSeeTableRecords(Course::get())
+        ->filterTable('tags',$vuesCourseTagsIds)
+        ->assertCanNotSeeTableRecords(Course::whereHas('tags',function ($q) use ($filamentCourseTagsIds){
+            $q->whereIn('tag_id',array_merge($filamentCourseTagsIds));
+        })->get())
+        ->assertSeeText([
+            'Vue',
+        ])->removeTableFilters()
+        ->assertSeeText(['Bootstrap'])
+        ->assertCanSeeTableRecords(Course::get());
+
+});
+
+
+it('can be reset the filter ', function () {
+
+    $courses = Course::factory(2)
+        ->has(Episode::factory())
+        ->has(
+            Tag::factory()
+                ->count(2)
+        )
+        ->create();
+
+    Livewire::test(CourseList::class)
+        ->assertCanSeeTableRecords($courses)
+        ->resetTableFilters();
+});
```

Comment:
I believe you are overcomplicating the test a little bit. 
First: You only really need one test case. So the 'can reset filter' and 'remove tag' test cases are unnecessary.
Second: The test case implementation can be reduced to the following
```php
it('can filter by tags', function () {
    // These records are just filler and don't require detailed tag names
    Course::factory(3)
        ->has(Episode::factory())
        ->has(Tag::factory()->count(3))
        ->create();
    // This record will be used to filter the record by its tag. Only one tag is needed
    // with a specific value
    $course = Course::factory()
        ->has(Episode::factory())
        ->has(
            Tag::factory()
                ->state(['name' => 'Laravel'])
        )
        ->create();
    
    Livewire::test(CourseList::class)
        // Make sure there are no exceptions
        ->assertOk()
        // Assert that you can see all records
        ->assertCanSeeTableRecords(Course::all())
        // Pass the tag ids you want to filter by
        ->filterTable('tags', $course->tags()->pluck('tag_id')->toArray())
        // Assert that you can see the records that belong to that tag
        ->assertCanSeeTableRecords(Course::whereRelation('tags', 'tag_id', $course->tags->first()->id)->get())
        // And finally, assert that you can't see any record with a different tag
        ->assertCanNotSeeTableRecords(Course::whereRelation('tags', 'tag_id', '!=', $course->tags->first()->id)->get());
});
```

Reply:
Thank for your review and I am happy to contribute . I have never write test case on my work because they don't give me enough time. But I knew when I learned your coding youtube test case is important to improve more clear and readable and maintainable. And , I will recorrect my test case below your suggestion.